crates/p2p/src/main_loop.rs

```diff
@@ -171,6 +178,49 @@ impl MainLoop {
             SwarmEvent::ConnectionEstablished {
                 peer_id, endpoint, ..
             } => {
+                // If this is an incoming connection, we have to prevent the peer from
+                // reconnecting too quickly.
+                if endpoint.is_listener() {
+                    // Extract the IP address of the peer from his multiaddr.
+                    let ip = endpoint.get_remote_address().iter().find_map(|p| match p {
+                        Protocol::Ip4(ip) => Some(IpAddr::V4(ip)),
+                        Protocol::Ip6(ip) => Some(IpAddr::V6(ip)),
+                        _ => None,
+                    });
+                    match ip {
+                        Some(ip) => {
+                            // If the peer is in the recent peers set, this means he is attempting to
+                            // reconnect too quickly. Close the connection.
+                            if self.recent_peers.read().await.contains(&ip) {
+                                tracing::debug!(%peer_id, "Peer attempted to reconnect too quickly, closing");
+                                if let Err(e) = self.disconnect(peer_id).await {
+                                    tracing::error!(%e, "Failed to disconnect peer");
+                                }
+                                return;
+                            } else {
+                                // Otherwise, add the peer to the recent peers set. After a certain period,
+                                // remove the peer from the recent peers set. This allows us to prevent
+                                // malicious peers from reconnecting too quickly.
+                                let recent_peers = self.recent_peers.clone();
+                                let timeout = self.connection_timeout;
+                                recent_peers.write().await.insert(ip);
+                                tokio::spawn(async move {
+                                    tokio::time::sleep(timeout).await;
+                                    recent_peers.write().await.remove(&ip);
+                                });
+                            }
```

Comment:
Consider moving parts of this into a `RecentPeers` type so that the logic for how this works moves out. I imagine this handler will get plenty large over time.

e.g. having 
```rust 
impl RecentPeers {
    fn insert(...) {...}
    fn contains(...) -> bool {...}
}
```

Reply:
+1